server/src/main/java/com/linecorp/centraldogma/server/internal/storage/ScheduledPurgeService.java

```diff
@@ -0,0 +1,234 @@
+import com.google.common.util.concurrent.MoreExecutors;
+import com.spotify.futures.CompletableFutures;
+
+import com.linecorp.centraldogma.common.Author;
+import com.linecorp.centraldogma.server.command.Command;
+import com.linecorp.centraldogma.server.command.CommandExecutor;
+import com.linecorp.centraldogma.server.metadata.MetadataService;
+import com.linecorp.centraldogma.server.metadata.UserAndTimestamp;
+import com.linecorp.centraldogma.server.storage.project.ProjectManager;
+
+import io.netty.util.concurrent.DefaultThreadFactory;
+
+/**
+ * A service class for pursing projects and repositories that were removed before.
+ */
+public class ScheduledPurgeService {
+
+    private static final Logger logger = LoggerFactory.getLogger(ScheduledPurgeService.class);
+
+    /**
+     * How often to run the storage purge schedules. i.e. every minute.
+     */
+    private static final Duration TICK = Duration.ofMinutes(1);
+    private static final Duration INITIAL_DELAY = Duration.ofSeconds(1);
+
+    private final ProjectManager projectManager;
+    private final long maxRemovedRepositoryAgeMillis;
+    private final StoragePurgeScheduler storagePurgeScheduler = new StoragePurgeScheduler();
+
+    public ScheduledPurgeService(
+            ProjectManager projectManager,
+            long maxRemovedRepositoryAgeMillis) {
+        this.projectManager = requireNonNull(projectManager, "projectManager");
+        checkArgument(maxRemovedRepositoryAgeMillis >= 0,
+                      "maxRemovedRepositoryAgeMillis: %s (expected: >= 0)", maxRemovedRepositoryAgeMillis);
+        this.maxRemovedRepositoryAgeMillis = maxRemovedRepositoryAgeMillis;
+    }
+
+    public void start(CommandExecutor commandExecutor,
+                      MetadataService metadataService) {
+        if (!isDisabled()) {
+            requireNonNull(commandExecutor, "commandExecutor");
+            requireNonNull(metadataService, "metadataService");
+            cleanPurgedFiles(commandExecutor);
+            storagePurgeScheduler.start(() -> schedulePurgeData(commandExecutor, metadataService).join());
+        }
+    }
+
+    public void stop() {
+        if (!isDisabled()) {
+            storagePurgeScheduler.stop();
+        }
+    }
+
+    @VisibleForTesting
+    void cleanPurgedFiles(CommandExecutor commandExecutor) {
+        projectManager.listPurged()
+                      .forEach(name -> commandExecutor.execute(Command.purgeProject(Author.SYSTEM, name)));
+        projectManager.list().values()
+                      .forEach(project -> project
+                              .repos()
+                              .listPurged()
+                              .forEach(repoName -> commandExecutor.execute(
+                                      Command.purgeRepository(Author.SYSTEM, project.name(), repoName))));
+    }
+
+    @VisibleForTesting
+    CompletableFuture<Void> schedulePurgeData(CommandExecutor commandExecutor,
```

Comment:
Is it necessary to make `schedulePurgeData`, `purgeProject` and `purgeRepository` asynchronously and concurrently at all? How about simplifying it to run synchronously and sequentially, for example just using `purgeWorker` we created when starting `CentralDogma`.

Reply:
.. and then we do not need to return `CompletableFuture` at all.

Reply:
I changed `schedulePurgeData` into synchronous and sequential logic. It doesn't return CompletableFuture anymore. :-)